test/Interop/Cxx/function/template-irgen.swift

```diff
@@ -0,0 +1,89 @@
+// RUN: %target-swift-emit-ir %s -I %S/Inputs -enable-cxx-interop | %FileCheck %s
+
+import FunctionTemplates
+
+// CHECK-LABEL: define {{.*}}i32 @"$s4main20testPassThroughConst1xs5Int32VAE_tF"(i32 %0)
+// CHECK: [[RET_VAL:%.*]] = call i32 @_Z16passThroughConstIiEKT_S0_(i32 %0)
+// CHECK: ret i32 [[RET_VAL]]
+
+// CHECK-LABEL: define linkonce_odr i32 @_Z16passThroughConstIiEKT_S0_(i32 %value)
+public func testPassThroughConst(x: Int32) -> Int32 {
+  return passThroughConst(x)
+}
+
+// CHECK-LABEL: define {{.*}}i32 @"$s4main15testPassThrough1xs5Int32VAE_tF"(i32 %0)
+// CHECK: [[RET_VAL:%.*]] = call i32 @_Z11passThroughIiET_S0_(i32 %0)
+// CHECK: ret i32 [[RET_VAL]]
+
+// CHECK-LABEL: define linkonce_odr i32 @_Z11passThroughIiET_S0_(i32 %value)
+public func testPassThrough(x: Int32) -> Int32 {
+  return passThrough(x)
+}
+
+// CHECK-LABEL: define {{.*}}i32 @"$s4main19testAddTwoTemplates1xs5Int32VAE_tF"(i32 %0)
+// CHECK: [[OUT_VAL:%.*]] = call i32 @_Z15addTwoTemplatesIiiET_S0_T0_(i32 %0, i32 %0)
+// CHECK: ret i32 [[OUT_VAL]]
+
+// CHECK-LABEL: define linkonce_odr i32 @_Z15addTwoTemplatesIiiET_S0_T0_(i32 %a, i32 %b)
+public func testAddTwoTemplates(x: Int32) -> Int32 {
+  return addTwoTemplates(x, x)
+}
+
+// CHECK-LABEL: define {{.*}}i32 @"$s4main7testAdd1xs5Int32VAE_tF"(i32 %0)
+// CHECK: [[OUT_VAL:%.*]] = call i32 @_Z3addIiET_S0_S0_(i32 %0, i32 %0)
+// CHECK: ret i32 [[OUT_VAL]]
+
+// CHECK-LABEL: define linkonce_odr i32 @_Z3addIiET_S0_S0_(i32 %a, i32 %b)
+public func testAdd(x: Int32) -> Int32 {
+  return add(x, x)
+}
+
+// TODO: Generics supplied as template parameters.
+// X-CHECK-LABEL: define {{.*}}void @"$s4main22testPassThroughGeneric1xxx_tlF"(%swift.opaque* noalias nocapture sret %0, %swift.opaque* noalias nocapture %1, %swift.type* %T)
+// X-CHECK-LABEL: define linkonce_odr i8* @_Z11passThroughIP11objc_objectET_S2_(i8* %value)
+// public func testPassThroughGeneric<T>(x: T) -> T {
```

Comment:
I think generic types should be a follow-up patch. The "issue" is that we need to generate a `FuncDecl` for the specialized clang function template with the specialized parameters. The specialized parameters are generic types so, they can't exist outside of the caller function. However, I think this can be solved pretty easily by just giving the new `FuncDecl` a generic parameter for each generic it gets from the caller.